src/app/services/task.service.spec.ts

```diff
@@ -0,0 +1,149 @@
+];
+
+describe('TaskService', () => {
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      providers: [
+        TaskService,
+        { provide: WebService, useClass: WebServiceMock }
+      ]
+    });
+    taskService = TestBed.get(TaskService);
+    webService = TestBed.get(WebService);
+    taskService.tasks = tasks;
+  });
+
+  it('should be created', () => {
+    expect(taskService).toBeTruthy();
+  });
+
+  it('call renderAllTasks() => do not call WebService.getTasks() if local storage return array of tasks', () => {
+    spyOn(localStorage, 'getItem').and.returnValue(JSON.stringify(tasks));
+    const spy = spyOn(webService, 'getTasks').and.callThrough();
+    taskService.renderAllTasks();
+    expect(spy).not.toHaveBeenCalled();
+  });
+
+  it('call renderAllTasks() => call WebService.getTasks() if local storage return empty array', () => {
+    spyOn(localStorage, 'getItem').and.returnValue('[]');
+    const spy = spyOn(webService, 'getTasks').and.callThrough();
+    taskService.renderAllTasks();
+    expect(spy).toHaveBeenCalled();
+  });
+
+  it('call renderAllTasks() => call WebService.getTasks() if local storage return null', () => {
+    spyOn(localStorage, 'getItem').and.returnValue(null);
+    const spy = spyOn(webService, 'getTasks').and.callThrough();
+    taskService.renderAllTasks();
+    expect(spy).toHaveBeenCalled();
+  });
+
+  it('call addTask() => call TaskService.updateLocalStorage()', () => {
+    const spy = spyOn(taskService, 'updateLocalStorage').and.stub();
+    taskService.addTask(task);
+    expect(spy).toHaveBeenCalled();
+  });
+
+  it('call addTask() => add addedTask to tasks', () => {
+    const addedTask = {name: 'rrrr', isCompleted: false};
+    taskService.addTask(addedTask);
+    expect(taskService.tasks).toContain(addedTask);
+  });
+
+  it('call toggleTask() => call TaskService.updateLocalStorage()', () => {
+    const spy = spyOn(taskService, 'updateLocalStorage').and.stub();
+    taskService.toggleTask(task);
+    expect(spy).toHaveBeenCalled();
+  });
+
+  it('call toggleTask() => toggle task.isCompleted', () => {
+    const status = task.isCompleted;
+    taskService.toggleTask(task);
+    expect(status).toBe(!task.isCompleted);
+  });
+
+  it('call toggleAllTask() => call TaskService.updateLocalStorage()', () => {
+    const spy = spyOn(taskService, 'updateLocalStorage').and.stub();
+    taskService.toggleAllTasks(true);
+    expect(spy).toHaveBeenCalled();
+  });
+
+  it('call toggleAllTask(true) => set all tasks isCompleted = true', () => {
+    taskService.toggleAllTasks(true);
+    const _tasks =  taskService.tasks.filter((_task: Task) => {
+      return _task.isCompleted === true;
+    });
+    const result = (_tasks.length === taskService.tasks.length);
+    expect(result).toBeTruthy();
+  });
+
+  it('call toggleAllTask(false) => set all tasks isCompleted = false', () => {
+    taskService.toggleAllTasks(false);
+    const _tasks =  taskService.tasks.filter((_task: Task) => {
```

Comment:
Make mocks customizable, try to avoid using logic inside tests

Reply:
Nope

Reply:
Ok